geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/command/ExecuteFunction66.java

```diff
@@ -228,6 +229,7 @@ public void cmdExecute(final Message clientMessage, final ServerConnection serve
       int earlierClientReadTimeout = handshake.getClientReadTimeout();
       handshake.setClientReadTimeout(functionTimeout);
 
+      long statStartTime = stats.getTime();
```

Comment:
It looks as if we're getting the start time only so that this method can update failure stats when `executeFunctionOnGroups()` and `executeFunctionLocally()` decide not to catch and handle exceptions themselves. Is there a way to push all of the exception handling into those methods, so that this method doesn't have to worry about updating the execution failure stats at all?

Reply:
I like the idea, but in my opinion this whole method and especially `executeFunctionLocally()` needs to be rewritten because there appear to be multiple problems with the way stats are being recorded. I can't think of a way to fix this issue without making significant changes to these methods which seems out of scope for this PR.

Reply:
> multiple problems with the way stats are being recorded

Do you mean that the stats have incorrect values?

If so, do our new meters also have those problems? If so, I think it's in scope. Or we need to renegotiate the scope.

If the meters are correct, then I agree that fixing problems in the stats is out of scope.

Reply:
> multiple problems with the way stats are being recorded

For example, `executeFunctionLocally()` will schedule a function execution on an executor and then immediately call `endFunctionExecution()` even though the scheduled function execution has not necessarily been run at this point. To me this looks wrong, but I haven't been able to test this case specifically.

> Do you mean that the stats have incorrect values?

We haven't tested this code path specifically, but we have tested all of the scenarios in the JIRA ticket and those all have the correct values.

Reply:
We dug into this method and it turned out that `executeFunctionLocally()` gets called when executing a function using `FunctionService.onServer()` or `FunctionService.onServers()`. In the case where the function does not return a result (`hasResult()` returns `false`), the recorded execution time will be incorrect because the code does not wait for the execution to complete before recording that the execution finished.

We fixed this issue in the latest commits and added additional acceptance tests to cover scenarios where we execute a function that does not return a result.

Reply:
I believe this issue is resolved, but please review the latest changes and add a review comment on the latest changes if you find any issue.